### PR TITLE
CATL-1303: Cases I Am Involved to also include activities

### PR DESCRIPTION
**Pull request by reneolivo**
## Overview
This PR changes the behaviour for the "Cases I am involved" filter. Depending on a setting, the filter will include cases where the user is involved in a case activity and not directly to the case.

### Data setup

**Users:**
* U1 = Logged in user
* U2 and U3 = other users.

**Cases:**
* C1: U1 is case manager
* C2: U1 is a role
* C3: U1 is a client
* C4: U1 created an activity, assigned to U2, targeted to U3
* C5: U2 created an activity, assigned to U3, targeted to U1
* C6: U3 created an activity, assigned to U1, targeted to U2
* C7: U2 is case manager
* C8: U2 is a role
* C9: U2 is a client

### Before
Manage_CTC1s_c8008 (2)

#### When the setting is ON
Manage_CTC1s_c8008

### Dashboard when the setting is On
gif

#### When the setting is OFF
Manage_CTC1s_c8008 (1)
Same as the previous behaviour

## Technical Details

* A new `includeActivitiesForInvolvedContact` setting has been added.
* The `CRM_Civicase_APIHelpers_CasesByContactInvolved` class was refactored so it can return the SQL queries for filtering cases by contact involved, and for activity contact involved. The `filter` was refactored so it uses the new helper functions internally, but it remains the same in terms of parameters and output because this function is used by other endpoints.
* A new `CRM_Civicase_APIHelpers_CaseDetails` class has been provided which contains most of the code used in `Case.Getdetails` plus the new `CRM_Civicase_APIHelpers_CasesByContactInvolved` implementation. The reason for moving this code to a helper class was so it can be used for both: `Case.Getdetails` and `Case.Getdetailscount`.
* The `Case.Getdetailscount` was refactored because it had a bug: when the same contact had multiple involvements to the same case it would include this case multiple times. For `Case.Getdetails` this was fixed by using a `group by a.id`, but grouping does not work for counting. For these scenarios where the same record is repeated multiple times and we only want to count unique records we use `DISTINCT`. The problem is that core does not support `DISTINCT`. To force this statement on the SELECT part of the query we created a new class that extends from `Api3SelectQuery`, but which allows to replace the `selectFields` property which allows us to add the DISTINCT statement. This is implemented in `_civicrm_api3_case_getdetailscount_get_count_query` which follows a similar pattern to [CiviCRM's basic get function](https://github.com/compucorp/civicrm-core/blob/master/api/v3/utils.php#L1239-L1262), but focuses solely on counting.
* The search and dashboard directives will pass the `has_activities_for_involved_contact` filter equal to `1` if the setting is ON. this will trigger the new behaviour.

### CODE REVIEW

File: CRM/Civicase/APIHelpers/CasesByContactInvolved.php [hunk 1]
```diff
@@ -16,15 +16,49 @@ class CRM_Civicase_APIHelpers_CasesByContactInvolved {
    *   The ID of the contact related to the case.
    */
   public static function filter(CRM_Utils_SQL_Select $query, $contactInvolved) {
```
Comment (tunbola) [inline CRM/Civicase/APIHelpers/CasesByContactInvolved.php:18]: Looks like this function is not being used again is it?

File: ang/civicase/dashboard/directives/dashboard-tab.directive.js [hunk 2]
```diff
@@ -46,9 +46,10 @@
         'case_id.contacts'
       ]
     };
+    // The ID is needed for the sort to avoid returning cases in a random order
     var CASES_QUERY_PARAMS_DEFAULTS = {
       'status_id.grouping': 'Opened',
-      options: { sort: 'start_date DESC' },
+      options: { sort: 'start_date DESC, id DESC' },
```
Comment (deb1990) [inline ang/civicase/dashboard/directives/dashboard-tab.directive.js:52]: If order was wrong in Dashboard, wont it affect Manage Cases too? Also did we find why the order is wrong?

Comment (reneolivo) [reply ang/civicase/dashboard/directives/dashboard-tab.directive.js:52]: Manage cases has something that always appends the `id` parameter to the query. I was searching if there was a note for this, but this was added by core without a note or PR:

https://github.com/compucorp/uk.co.compucorp.civicase/commit/211cdf33e3fdfbb1bc43a236605b1cfa56256cb5#diff-73e074d6c90a338205dbee2d416c987dR277

File: api/v3/Case/Getstats.php [hunk 3]
```diff
@@ -74,7 +80,11 @@ function civicrm_api3_case_getstats(array $params) {
   }
 
   if (!empty($params['contact_involved'])) {
-    CRM_Civicase_APIHelpers_CasesByContactInvolved::filter($query, $params['contact_involved']);
+    CasesByContactInvolved::filter(
+      $query,
+      $params['contact_involved'],
+      $params['has_activities_for_involved_contact']
```
Comment (tunbola) [inline api/v3/Case/Getstats.php:86]: Was this variable initialised somewhere?
If not should be something like 
```php
isset($params['has_activities_for_involved_contact']) ? $params['has_activities_for_involved_contact'] : NULL
```

**Pull request merged**